File: sdks/python/src/verdocs/resources/webhooks.py
"""Webhook operations (js-sdk: Organizations/Webhooks.ts).

An organization has at most one webhook configuration. It cannot be deleted;
disable it by setting active to False or the URL to an empty string. Every
call here requires an admin or owner.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..models.core import Webhook
from ..models.organizations import WebhookSetParams

if TYPE_CHECKING:
    from .._endpoint import AsyncVerdocsEndpoint, VerdocsEndpoint

_WEBHOOKS_PATH = "/v2/webhooks"


def _write_body(params: WebhookSetParams) -> dict[str, Any]:
    return params.model_dump(mode="json", exclude_unset=True)


class Webhooks:
    """Webhook calls for a sync endpoint."""

    def __init__(self, endpoint: VerdocsEndpoint) -> None:
        self._endpoint = endpoint

    def get(self) -> Webhook:
        """Get the webhook configuration for the caller's organization via GET /v2/webhooks.

        The secret_key and client_secret come back masked to their last four
        characters; the full secret_key is only shown by rotate_secret().
        Mirrors js-sdk getWebhooks.

        Returns:
            The organization's webhook configuration.

        Raises:
            NotFoundError: The organization has never configured a webhook.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("GET", _WEBHOOKS_PATH)
        return Webhook.model_validate(response.json())

    def set(self, params: WebhookSetParams) -> Webhook:
        """Update (or first configure) the organization's webhook via PATCH /v2/webhooks.

        Mirrors js-sdk setWebhooks.

        Example:
            webhook = endpoint.webhooks.set(
                WebhookSetParams(
                    url="https://example.com/hooks/verdocs",
                    active=True,
                    events={"envelope_created": True, "envelope_completed": True},
                )
            )

        Args:
            params: The full configuration to apply. The URL must be HTTPS,
                or "" to disable deliveries.

        Returns:
            The updated webhook configuration.

        Raises:
            VerdocsAPIError: The API rejected the request.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("PATCH", _WEBHOOKS_PATH, json=_write_body(params))
        return Webhook.model_validate(response.json())

    def rotate_secret(self) -> Webhook:
        """Rotate (or first create) the webhook signing secret via PUT /v2/webhooks/rotate-secret.

        Until a secret exists, webhook calls carry no signature header. To
        authenticate a delivery, HMAC-SHA256 the JSON of the payload's inner
        "body" field with the secret and compare against the
        x-webhook-signature header. Pending deliveries keep the old secret
        until the next webhook fires. Mirrors js-sdk rotateWebhookSecret.

        Example:
            webhook = endpoint.webhooks.rotate_secret()
            signing_secret = webhook.secret_key  # unmasked only in this response

        Returns:
            The webhook configuration including the new secret_key.

        Raises:
            NotFoundError: The organization has never configured a webhook.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("PUT", f"{_WEBHOOKS_PATH}/rotate-secret")
        return Webhook.model_validate(response.json())


class AsyncWebhooks:
    """Webhook calls for an async endpoint."""

    def __init__(self, endpoint: AsyncVerdocsEndpoint) -> None:
        self._endpoint = endpoint

    async def get(self) -> Webhook:
        """Get the webhook configuration for the caller's organization via GET /v2/webhooks.

        The secret_key and client_secret come back masked to their last four
        characters; the full secret_key is only shown by rotate_secret().
        Mirrors js-sdk getWebhooks.

        Returns:
            The organization's webhook configuration.

        Raises:
            NotFoundError: The organization has never configured a webhook.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", _WEBHOOKS_PATH)
        return Webhook.model_validate(response.json())

    async def set(self, params: WebhookSetParams) -> Webhook:
        """Update (or first configure) the organization's webhook via PATCH /v2/webhooks.

        Mirrors js-sdk setWebhooks.

        Example:
            webhook = await endpoint.webhooks.set(
                WebhookSetParams(
                    url="https://example.com/hooks/verdocs",
                    active=True,
                    events={"envelope_created": True, "envelope_completed": True},
                )
            )

        Args:
            params: The full configuration to apply. The URL must be HTTPS,
                or "" to disable deliveries.

        Returns:
            The updated webhook configuration.

        Raises:
            VerdocsAPIError: The API rejected the request.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("PATCH", _WEBHOOKS_PATH, json=_write_body(params))
        return Webhook.model_validate(response.json())

    async def rotate_secret(self) -> Webhook:
        """Rotate (or first create) the webhook signing secret via PUT /v2/webhooks/rotate-secret.

        Until a secret exists, webhook calls carry no signature header. To
        authenticate a delivery, HMAC-SHA256 the JSON of the payload's inner
        "body" field with the secret and compare against the
        x-webhook-signature header. Pending deliveries keep the old secret
        until the next webhook fires. Mirrors js-sdk rotateWebhookSecret.

        Example:
            webhook = await endpoint.webhooks.rotate_secret()
            signing_secret = webhook.secret_key  # unmasked only in this response

        Returns:
            The webhook configuration including the new secret_key.

        Raises:
            NotFoundError: The organization has never configured a webhook.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("PUT", f"{_WEBHOOKS_PATH}/rotate-secret")
        return Webhook.model_validate(response.json())

File: sdks/csharp/src/Verdocs.Sdk/Resources/Webhooks.cs
using Verdocs.Models;

namespace Verdocs.Resources;

/// <summary>
/// Webhook calls, reached through the endpoint's Webhooks property. An organization has one
/// webhook configuration that delivers event notifications to a caller-supplied URL; every
/// call requires the caller to be an admin.
/// </summary>
public sealed class Webhooks
{
    private readonly VerdocsEndpoint _endpoint;

    internal Webhooks(VerdocsEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    /// <summary>
    /// Gets the webhook configuration for the caller's organization. The server answers 404
    /// when no configuration has ever been set. The client secret and signing secret are
    /// masked to their last four characters in this response.
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The organization's webhook configuration.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because nothing is configured yet.</exception>
    /// <sdkOperation>webhook.getWebhooks</sdkOperation>
    /// <sdkGroup>Webhook</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Webhook> GetAsync(CancellationToken cancellationToken = default)
    {
        return _endpoint.SendAsync<Webhook>(HttpMethod.Get, "/v2/webhooks", null, cancellationToken);
    }

    /// <summary>
    /// Sets the webhook configuration for the caller's organization. Configurations cannot be
    /// deleted; disable deliveries by setting Active to false or the URL to an empty string.
    ///
    /// <example>
    /// <code>
    /// await endpoint.Webhooks.SetAsync(new SetWebhookRequest
    /// {
    ///     Url = "https://hooks.example.com/verdocs",
    ///     Active = true,
    ///     AuthMethod = WebhookAuthMethod.Hmac,
    ///     Events = new Dictionary&lt;string, bool&gt; { [WebhookEvent.EnvelopeCompleted] = true },
    /// });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="request">The configuration to apply.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The updated webhook configuration.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the URL is not HTTPS.</exception>
    /// <sdkOperation>webhook.setWebhooks</sdkOperation>
    /// <sdkGroup>Webhook</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Webhook> SetAsync(SetWebhookRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<Webhook>(HttpMethod.Patch, "/v2/webhooks", request, cancellationToken);
    }

    /// <summary>
    /// Rotates (or first creates) the secret used to sign webhook deliveries. Until a secret
    /// exists, deliveries carry no signature header. Deliveries already queued keep the old
    /// signature; new events use the new secret. To authenticate a delivery, compare an
    /// HMAC-SHA256 hex digest of the payload's inner "body" field against the
    /// x-webhook-signature request header.
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The webhook configuration including the new, unmasked secret key.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because nothing is configured yet.</exception>
    /// <sdkOperation>webhook.rotateWebhookSecret</sdkOperation>
    /// <sdkGroup>Webhook</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Webhook> RotateSecretAsync(CancellationToken cancellationToken = default)
    {
        return _endpoint.SendAsync<Webhook>(HttpMethod.Put, "/v2/webhooks/rotate-secret", null, cancellationToken);
    }
}
